### PR TITLE
e2e: provision cluster entirely through Terraform

### PR DESCRIPTION
**Pull request by tgross**
Fixes https://github.com/hashicorp/nomad/issues/7069 https://github.com/hashicorp/nomad/issues/7068 (and some private repo issues)

Have Terraform run the target-specific `provision.sh`/`provision.ps1` script
rather than the test runner code which needs to be customized for each
distro. Use Terraform's detection of variable value changes so that we can
re-run the provisioning without having to re-install Nomad on those specific
hosts that need it changed.

Allow the configuration "profile" (well-known directory) to be set by a
Terraform variable. The default configurations are installed during Packer
build time, and symlinked into the live configuration directory by the
provision script. Detect changes in the file contents so that we only upload
custom configuration files that have changed between Terraform runs.

---

Note for reviewers:
* This looks like a beast of a PR but a big chunk of it is cleaning out now-unused code. Best reviewed commit-by-commit, but the "meat" of it is in the first commit.
* I want to hold off merging this until after 0.12.4 ships so that we're not interrupting testing efforts. I've given this some testing but I'll use it as the basis for my own testing for 0.12.4 so that it gets a full workout before I foist it on the rest of the development team.

### CODE REVIEW

**Comment by tgross** [comment]
(I'm converting this back to a draft until I've got an issue around upgrades figured out.)

**Comment by tgross** [comment]
I've got this mostly working (minus some quirks getting Consul running as a service on Windows), but after a good number of runs I'm realizing this approach is still pushing too much configuration to runtime rather than AMI build-time, and in the common case of the full-cluster or dev-cluster profile we can get this down to just one invocation of remote-exec per instance. Going to rework this slightly to allow for that.

**Comment by tgross** [comment]
I've now got this working but the size of this PR is absurd... I'm going to see if I can pull out some reviewable bits and maybe make separate `dev-cluster` config profile something that gets done later.

**Comment by tgross** [comment]
@schmichael and @notnoop I managed to pull about half the code in the original draft out to the now-merged #8821, #8823, #8824, #8827, and #8828.

* 9452f48 is kind of a big commit but is all test code to verify the Terraform logic without having to stand up a whole cluster to do it. I'm not 100% convinced we should even keep it once we've completed the E2E work we have planned for the next month or so, but it'll be handy until then.
* 346e08a is ditching all the old framework code that this new code replaces.

I've used this to stand up a working cluster and I'm happy with the result, but we'll need to drop e441c0d and rebase before merging.

**Comment by tgross** [comment]
@nickethier @notnoop @schmichael in the interest of trying to make this a little less intimidating to review, I've pulled out https://github.com/hashicorp/nomad/pull/8907 and https://github.com/hashicorp/nomad/pull/8908 as separate draft PRs (they were their own commits here previously).

I've used this to stand up a working cluster and I'm happy with the result, but we'll need to update TeamCity (I've got a PR in draft in `nomad-e2e`) before merging.

File: e2e/terraform/packer/linux/provision.sh [hunk 1]
```diff
@@ -59,14 +65,46 @@ install_from_release() {
...
+install_config_profile() {
+
+    if [ -d /tmp/custom ]; then
+        rm -rf /opt/config/custom
+        sudo mv /tmp/custom /opt/config/
+    fi
+
+    sudo rm -rf /etc/nomad.d/*
+    sudo rm -rf /etc/consul.d/*
+    sudo rm -rf /etc/vault.d/*
```
Comment (notnoop) [inline e2e/terraform/packer/linux/provision.sh:89]: I'd suggest deleting the folder and mkdir them again.  `*` has its own quarks (e.g. dot files, large dirs).

File: e2e/terraform/README.md [hunk 2]
```diff
@@ -37,6 +33,67 @@ cd e2e/terraform/
...
+* `custom`: This profile is used for one-off developer testing of more complex
+  interactions between features. You can build your own custom profile by
+  writing config files to the `./config/custom` directory, which are protected
+  by `.gitignore`
+
+For each profile, application (Nomad, Consul, Vault), and agent type
+(`server`, `client_linux`, or `client_windows`), the agent gets the following
+configuration files, ignoring any that are missing.
+
+* `./config/<profile>/<application>/*`: base configurations shared between all
```
Comment (notnoop) [inline e2e/terraform/README.md:81]: I was a bit thrown off by not seeing `config/profile` changes but I see it was merged in https://github.com/hashicorp/nomad/pull/8828

**Review by notnoop — APPROVED** [review]
Looks pretty good.  Love the concept of profiles and I'm happy to see one step gone.  Thanks!

File: e2e/terraform/packer/linux/provision.sh (hunk 1 above)
Comment (tgross) [reply e2e/terraform/packer/linux/provision.sh:89]: Good catch. Will fix.

**Pull request merged**